### ISSUE TITLE
add export capability

### ISSUE THREAD

**Issue by pcause** [mtobeiyf/pile#15]
I know it must be on your list, but being able to export a board (all sections) as a pdf file or as a single markdown document would be great.  I suspect downing the markdown export will be the easiest and then we can use something like pandoc to convert to something else.  Actually, creating a markdown document and then having the ability to send to pandoc, if installed, or other program would be even more awesome.

**Comment by mtobeiyf** [comment]
@pcause  Hi, thanks.
It's a nice feature to have the page exported to other formats. Electron has built-in API, kinda like the print-to-pdf you used in Chrome/Firefox. It's easy to add.
But I'm still considering the details of the exported information and how the exported document means to our work. Could you describe a detailed scene where the exported sections/markdown could be further exploited?
Currently, the note and todo sections are reasonable to be exported as markdown.

**Comment by mtobeiyf** [comment]
Thanks. 
I get your point~ 😃 
The exporting functionality will be added in the future releases.

And if you have more ideas on this, just comment on this issue.

**Comment by pcause** [comment]
one minor tweak to my idea - allow a set of boards to be exported to a single markdown document.  example : each board is a chapter of a report. or each board might be data about a feature of a set of features for a sprint being checked into a code tree.

**Comment by pcause** [comment]
another one (sorry, but there is such a great base here): export the json for a board(s) and allow it to be imported into another Board.

**Comment by mtobeiyf** [comment]
I'll implement exporting one board first. :sweat_smile: You can append them manually.
Export&import boards is much easier. Perhaps the json step could be skipped, just do the moving/copying inside the app.

**Comment by pcause** [comment]
wonder if it might be a good idea to have each board be a separate json file.  they can each be named with the board name and on startup you read the db files and create tab names from file names, when you open a tab you read the file.  when you switch tabs you write out state of board being left.

this allows for boards to be larger and also avoids risk of an issue wiping out all boards.

might move these into a subdirectory of the pile store location.

**Comment by mtobeiyf** [comment]
The intuition is good. 
However, it would be difficult to determine the order of these separate json files.
Currently, the size of the stored data is limited. And you can easily back up by copying the db json file.

**Comment by pcause** [comment]
2 options:
- have a settings.json.  users defined order plus other overall settings go here.

- use the last modified time to define the order.

**Comment by mtobeiyf** [comment]
Found something interesting https://www.notion.so/product 😃

**Comment by pcause** [comment]
yes, it is.  but I am liking Pile and see potential, particularly as you are open source.

one other thought on multiple files.  How about just a "file Open/Save/Save As" capability.  Use the same json just let the user pick which to open and then save/save as.  This should be pretty easy.    The issue is which to open on startup and you can likely use last modified time or put something in preferences file.  BTW, what is the file size limit?

**Comment by pcause** [comment]
actually, another thought is perhaps to use https://pouchdb.com/ to store the data.  This can be a local database, but then provides the future capability to sync to a cloud based clouchdb without a huge amount of work.  that would be really interesting.

**Comment by pcause** [comment]
here is a guide to using in electron https://github.com/nolanlawson/pouchdb-electron

**Comment by mtobeiyf** [comment]
"Open/Save/Save As" sounds good. I think this function should be called "Import/Export", which can save and restore the whole current workspace (tabs).  This function could be realized with your #12 by specifying different data file paths. This could be done soon. 😄 

Speaking of syncing, it's always good to have all the data synced between devices or keep it safe.
Actually, pile was not intended to be designed for note-taking or similar purposes like [notion](https://www.notion.so/product). There has already been enough great apps to do these specific tasks (Microsoft Todo, Evernote, OneNote, Trello...) It was designed as a management/productivity tool. 🔨  Making the data safe is very important, certainly. It is more convenient for users to have the data synced using their own daily cloud storage services like Dropbox, Google Drive, One Drive... ☁️ 

The db library looks friendly. However, it would be too time-consuming to switch to another db controller.  🤔 
Currently, lowdb can satisfy all the requirements of this app. 
Once json-based db cannot handle the needs, switching db would be inevitable.

BTW, if you'd like, you can contribute to this project.

**Comment by pcause** [comment]
I've looked at and/or used a lot of these.  Pile is the closest to what I'd love to see.  And as for productivity, I use multiple computers and a smartphone.  My work laptop is Widnows and my home is linux.  So sync'ing is important to me.  Having an option to specify location to store things plus the import/export lets me use one of many different tools to sync.

I would love to contribute but I am rusty but will give it a try.  I've been more on the management/product management side than coding for a bit and so am rusty.  If I can what I'd like to add is making the todo's a bit richer.  There is a todo.txt format which is pretty simple and could be added I think.  It supports priorities and many have extended with simple date spec for due date in the form of due:mm-dd-yyy.   Rather than reorder in the board the todo items by date or priority, I'd make a window appear to do the display and ordering.
